bindings/go/blob/filesystem/blob_from_path.go

```diff
@@ -0,0 +1,313 @@
+
+	// Create a ReadOnlyBlob from the TAR archive.
+	var tarBlob blob.ReadOnlyBlob = direct.New(reader, direct.WithMediaType(opt.mediaType()))
+	if opt.Compress {
+		tarBlob = compression.Compress(tarBlob)
+	}
+	return tarBlob, nil
+}
+
+// createTarStream creates a TAR archive from the contents of the specified path, either a directory or a single file.
+func createTarStream(ctx context.Context, path string, opt *DirOptions) (io.Reader, error) {
+	fi, err := os.Stat(path)
+	if err != nil {
+		return nil, fmt.Errorf("error stating path %q: %w", path, err)
+	}
+
+	var baseDir, subPath string
+
+	// Directory case
+	if fi.IsDir() {
+		if !opt.PreserveDir {
+			baseDir = path
+			subPath = "."
+			// add parent directory if PreserveDir is set
+		} else {
+			baseDir = filepath.Dir(path)
+			subPath = filepath.Base(path)
+		}
+		// Single file case
+	} else {
+		baseDir = filepath.Dir(path)
+		subPath = filepath.Base(path)
+	}
+
+	// Create a virtual filesystem rooted at the base directory.
+	fileSystem, err := NewFS(baseDir, os.O_RDONLY)
+	if err != nil {
+		return nil, fmt.Errorf("error creating virtual filesystem for path %q: %w", baseDir, err)
+	}
+
+	// Create a TAR archive from the directory contents.
+	// We differentiate between directory and single file cases.
+	pr, pw := io.Pipe()
+
+	go func() {
+		tw := tar.NewWriter(pw)
+
+		// use local gerr to capture errors from tar creation
+		var gerr error
+
+		// take care to close tar writer and pipe and capture any error
+		defer func() { gerr = errors.Join(gerr, tw.Close()) }()
+		defer func() { _ = pw.CloseWithError(gerr) }()
+
+		if fi.IsDir() {
+			gerr = createTarFromDir(ctx, fileSystem, subPath, opt, tw)
+		} else {
+			gerr = createTarFromSingleFile(ctx, fileSystem, subPath, opt, tw)
+		}
+	}()
+	return pr, nil
+}
+
+// createTarFromDir writes the contents of a directory to the TAR archive.
+// subPath is the relative path within the virtual filesystem.
+func createTarFromDir(ctx context.Context, fileSystem FileSystem, subPath string, opt *DirOptions, tw *tar.Writer) error {
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("context cancelled while processing directory %q: %w", subPath, ctx.Err())
+	default:
+	}
+
+	dirEntries, err := fileSystem.ReadDir(subPath)
+	if err != nil {
+		return fmt.Errorf("error reading directory %q: %w", subPath, err)
+	}
+
+	for _, entry := range dirEntries {
+		ei, err := entry.Info()
+		if err != nil {
+			return fmt.Errorf("error getting file info for entry %q in directory %q: %w", entry.Name(), subPath, err)
+		}
+
+		// fail early for symlinks
+		if (ei.Mode() & fs.ModeSymlink) != 0 {
+			return fmt.Errorf("symlinks are not supported: found symlink %q in directory %q", entry.Name(), subPath)
+		}
+
+		// reconstruct the full path of the entry
+		entryPath := filepath.Join(subPath, entry.Name())
+		// Check if path is included based on include/exclude patterns from options.
+		ok, err := isPathIncluded(entryPath, opt.IncludeFiles, opt.ExcludeFiles)
```

Comment:
could be merged into an

`if ok, err := isPathIncluded(entryPath, opt.IncludeFiles, opt.ExcludeFiles); err != nil {} else !ok {} `